### ISSUE TITLE
run a function or procedure instead of jamf policy

### ISSUE THREAD

**Issue by mbracco1973** [dan-snelson/dialog-scripts#22]
Not everyone is using jamf pro (we do), but it would be cool to be able to run a function or procedure instead of running a policy triggered. So i replaced run_jamf_trigger "$trigger" with ${trigger} and created functions called ${trigger} eg. filveault(){ echo "filevault"; touch /tmp/filevault.installed} and checked on that filevault.installed file created inside that function to get the status installed.
A flag/script parameter to choose between jamf / localinstall to make the difference between run_jamf_trigger and ${trigger} would be the parameter to distinct between running a policy or local function.
Cleaning out those *.installed files in /tmp at begin of script to be sure to run clean

**Comment by dan-snelson** [comment]
Thanks for the feature request, @mbracco1973.

Will you please share a sample of the modifications you've implemented?

**Comment by dan-snelson** [comment]
@mbracco1973 A quick reminder that I’m waiting on you for additional information; please above. Thanks.

**Comment by mbracco1973** [comment]
# At the beginning we do a cleaning.....
rm -f /Users/Shared/filevault.done
rm -f /Users/Shared/sophosendpoint.done
rm -f /Users/Shared/office.done
rm -f /Users/Shared/printer.done

# In functions and procedures part

filevault(){
echo "FileVault running"
sleep 5
dialogUpdateSetupYourMac "progresstext: Doing Filevault stuff now"
sleep 10
touch /Users/Shared/filevault.done
}

sophosEndpoint(){
echo "Sophos Endpoint running"
touch /Users/Shared/sophosendpoint.done
}

microsoftOffice(){
echo "Microsoft Office running"
touch /Users/Shared/office.done
}

printerDriver(){
echo "Printer Driver running"
touch /Users/Shared/printer.done
}

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# This for loop will iterate over each distinct step in the policy_array array
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

for (( i=0; i<dialog_step_length; i++ )); do

    # Increment the progress bar
    dialogUpdateSetupYourMac "progress: $(( i * ( 100 / progress_total ) ))"

    # Creating initial variables
    listitem=$(get_json_value "${policy_array[*]}" "steps[$i].listitem")
    icon=$(get_json_value "${policy_array[*]}" "steps[$i].icon")
    progresstext=$(get_json_value "${policy_array[*]}" "steps[$i].progresstext")

    trigger_list_length=$(get_json_value "${policy_array[*]}" "steps[$i].trigger_list.length")

    # If there's a value in the variable, update running swiftDialog
    if [[ -n "$listitem" ]]; then dialogUpdateSetupYourMac "listitem: index: $i, status: wait, statustext: Installing …, "; fi
    if [[ -n "$icon" ]]; then dialogUpdateSetupYourMac "icon: ${setupYourMacPolicyArrayIconPrefixUrl}${icon}"; fi
    if [[ -n "$progresstext" ]]; then dialogUpdateSetupYourMac "progresstext: $progresstext"; fi
    if [[ -n "$trigger_list_length" ]]; then
        for (( j=0; j<trigger_list_length; j++ )); do

            # Setting variables within the trigger_list
            trigger=$(get_json_value "${policy_array[*]}" "steps[$i].trigger_list[$j].trigger")
            path=$(get_json_value "${policy_array[*]}" "steps[$i].trigger_list[$j].path")

            # If the path variable has a value, check if that path exists on disk
            if [[ -f "$path" ]]; then
                updateScriptLog "SETUP YOUR MAC DIALOG: INFO: $path exists, moving on"
                if [[ "${debugMode}" == "true" ]]; then sleep 0.5; fi
            else
                echo "trigger : ${trigger}"
# Here we modify maybe add a case or if
if [ policyorlocalparamter != jamf ]
then
                ${trigger}
else
                run_jamf_trigger "$trigger"
fi
            fi
        done
    fi

    # Validate the expected path exists
    updateScriptLog "SETUP YOUR MAC DIALOG: Testing for \"$path\" …"
    if [[ -f "$path" ]] || [[ -z "$path" ]]; then
        dialogUpdateSetupYourMac "listitem: index: $i, status: success, statustext: Installed"
        if [[ "$trigger" == "recon" ]]; then
            dialogUpdateSetupYourMac "listitem: index: $i, status: success, statustext: Updated"
        fi
    else
        dialogUpdateSetupYourMac "listitem: index: $i, status: fail, statustext: Failed"
        jamfProPolicyTriggerFailure="failed"
        exitCode="1"
        jamfProPolicyPolicyNameFailures+="• $listitem  \n"
    fi

done

**Comment by dan-snelson** [comment]
Thanks for the additional details, @mbracco1973; please allow me some time to review.

**Comment by dan-snelson** [comment]
@mbracco1973:

I'm struggling to see the benefit of checking for arbitrary files created in `/Users/Shared/` vs. the current behavior of checking for vendor-created files, which are written in known, established locations:

|     **Item**     |                                       **Current**                                      |             **Proposed**            |
|:----------------:|:--------------------------------------------------------------------------------------:|:-----------------------------------:|
| FileVault        | Line No. 324: `"path": "/Library/Preferences/com.apple.fdesetup.plist"`                | `/Users/Shared/filevault.done`      |
| Sophos Endpoint  | Line No. 335: `"path": "/Applications/Sophos/Sophos Endpoint.app/Contents/Info.plist"` | `/Users/Shared/sophosendpoint.done` |
| Microsoft Office | Line No. N/A: `"path": "/Applications/Microsoft Outlook.app/Contents/Info.plist`       | `/Users/Shared/office.done`         |

(Also, what's to prevent a local administrator from easily stumbling across and deleting the files you're writing to `/Users/Shared` ?)

---

I am, however, intrigued by the following:
- `echo "Sophos Endpoint running"`
- `echo "Microsoft Office running"`
- `echo "Printer Driver running"`

What criteria are you using to confirm that each of these items are **running** vs. simply _installed_?

**Comment by dan-snelson** [comment]
@mbracco1973:

Will you please also review [Issue No. 21](https://github.com/dan-snelson/dialog-scripts/issues/21)? (I'm wondering if Issue Nos. 21 and 22 should be combined.)

**Comment by mbracco1973** [comment]
We can combine those. The echos are just cosmetics. Can be ignored. The main goal was to have a possibility to run a function instead of a jamf trigger. The file path was just an example to control the process of dialog to be successful, running, failed......

**Comment by dan-snelson** [comment]
Thanks for confirming, @mbracco1973.

Closing Issue No. 22 as a duplicate of [Issue No. 21](https://github.com/dan-snelson/dialog-scripts/issues/21).
